src/lib/battery/battery.h

```diff
@@ -31,91 +31,157 @@
+ * This was decided against because the newer ModuleParams API provides more type safety and avoids code duplication.
+ *
+ * To add a third battery, follow these steps:
+ *  - Copy/Paste all of Battery2 to make Battery3
+ *  - Change all "BAT2_*" parameters to "BAT3_*" in Battery3
+ *  - Copy the file "battery_params_2.c" to "battery_params_3.c"
+ *  - Change all of the "BAT2_*" params in "battery_params_3.c" to "BAT3_*"
+ * This is not done now because there is not yet any demand for a third battery, and adding parameters uses up space.
  */
 
-#pragma once
-
-#include <uORB/topics/battery_status.h>
-#include <drivers/drv_hrt.h>
-#include <px4_module_params.h>
-
-
-class Battery : public ModuleParams
+/**
+ * Battery1 represents a battery calibrated by BAT1_* parameters.
+ */
+class Battery1 : public BatteryBase
 {
 public:
-	Battery();
+	Battery1();
 
 	/**
-	 * Reset all battery stats and report invalid/nothing.
+	 * This function migrates the old deprecated parameters like BAT_N_CELLS to the new parameters like BAT1_N_CELLS.
+	 * It checks if the old parameter is non-defaulT AND the new parameter IS default, and if so, it:
+	 *  - Issues a warning using PX4_WARN(...)
+	 *  - Copies the value of the old parameter over to the new parameter
+	 *  - Resets the old parameter to its default
+	 *
+	 * The 'name' parameter should be only the part of the parameter name that comes after "BAT1_" or "BAT_". It is
+	 * used only for the warning message. For example, for parameter BAT1_N_CELLS, name should be "N_CELLS".
+	 * (See the implementation of this function for why I have taken this strange choice)
+	 *
+	 * In an ideal world, this function would be protected so that only child classes of Battery1 could access it.
+	 * However, the way ModuleParams works makes it very difficult to inherit from a ModuleParams class.
+	 * For example, the AnalogBattery class in the Sensors module does not inherit this class; it just contains
+	 * a Battery1 member variable.
+	 *
+	 * The templating is complicated because every parameter is technically a different type. However, in normal
+	 * use, the template can just be ignored. See the implementation of Battery1::Battery1() for example usage.
+	 *
+	 * @tparam P1 Type of the first parameter
+	 * @tparam P2 Type of the second parameter
+	 * @tparam T Data type for the default value
+	 * @param oldParam Reference to the old parameter, as a ParamFloat<...>, ParamInt<...>, or ParamBool<...>
+	 * @param newParam Reference to the new paramater, as a ParamFloat<...>, ParamInt<...>, or ParamBool<...>
+	 * @param name The name of the parameter, WITHOUT the "BAT_" or "BAT1_" prefix. This is used only for logging.
+	 * @param defaultValue Default value of the parameter, as specified in PARAM_DEFINE_*(...)
 	 */
-	void reset(battery_status_s *battery_status);
+	template<class P1, class P2, typename T> static void
+	migrateParam(P1 &oldParam, P2 &newParam, const char *name, T defaultValue)
```

Comment:
Let's think about making this a generic parameter migration mechanism system wide.
@julianoes @davids5

Reply:
If this happens let me know - sounds like a feature to document.
Will feature need to migrate parameters by pattern? ie BATTERY_n_CONFIG

Reply:
@dagar we can pull this out later, I'd say.